core/src/ModelArray.cpp

```diff
@@ -0,0 +1,332 @@
+    copy.m_data = m_data + x;
+    return copy;
+}
+
+ModelArray ModelArray::operator-(const double& x) const
+{
+    ModelArray copy(type, m_name + "-" + std::to_string(x));
+    copy.m_data = m_data - x;
+    return copy;
+}
+
+ModelArray ModelArray::operator*(const double& x) const
+{
+    ModelArray copy(type, m_name + "*" + std::to_string(x));
+    copy.m_data = m_data * x;
+    return copy;
+}
+
+ModelArray ModelArray::operator/(const double& x) const
+{
+    ModelArray copy(type, m_name + "/" + std::to_string(x));
+    copy.m_data = m_data / x;
+    return copy;
+}
+
+ModelArray operator+(const double& x, const ModelArray& y) { return y + x; }
+
+ModelArray operator-(const double& x, const ModelArray& y) { return -(y - x); }
+
+ModelArray operator*(const double& x, const ModelArray& y) { return y * x; }
+
+ModelArray operator/(const double& x, const ModelArray& y)
+{
+    ModelArray xArray(y.getType(), std::to_string(x));
+    xArray.setData(x);
+    return xArray / y;
+}
+
+void ModelArray::setData(double value)
+{
+    resize();
+    m_data = value;
+}
+
+void ModelArray::setData(const double* pData)
+{
+    resize();
+    auto out = std::copy(pData, pData + m_sz.at(type), m_data.data());
+}
+
+void ModelArray::setData(const DataType& from) { setData(from.data()); }
+
+void ModelArray::setData(const ModelArray& from) { setData(from.m_data.data()); }
+
+void ModelArray::setDimensions(Type type, const Dimensions& newDims)
+{
+    size_t newSize = 1;
+    for (size_t dimLen : newDims) {
+        newSize *= dimLen;
+    }
+    m_dims.at(type).clear();
+    std::copy(newDims.begin(), newDims.end(), std::back_inserter(m_dims.at(type)));
+    m_sz.at(type) = newSize;
+}
+
+// Simple 1d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i) { return i; }
+
+// Special case for 2d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i, size_t j)
+{
+    return i * dims[1] + j;
+}
+
+// General case for up to 8d indexing
+inline size_t indexr(const size_t* dims, const size_t nDims, size_t i, size_t j, size_t k,
+    size_t l = 0, size_t m = 0, size_t n = 0, size_t p = 0, size_t q = 0)
+{
+    size_t stepLength = 1;
+    size_t out = 0;
+    switch (nDims) {
+    case 8:
+        out += stepLength * q;
+        stepLength *= dims[7];
+    case 7:
+        out += stepLength * p;
+        stepLength *= dims[6];
+    case 6:
+        out += stepLength * n;
+        stepLength *= dims[5];
+    case 5:
+        out += stepLength * m;
+        stepLength *= dims[4];
+    case 4:
+        out += stepLength * l;
+        stepLength *= dims[3];
+    case 3:
+        out += stepLength * k;
+        stepLength *= dims[2];
+    case 2:
+        out += stepLength * j;
+        stepLength *= dims[1];
+    case 1:
+    default:
+        out += stepLength * i;
+        return out;
+    }
+    return 0;
+}
+
+size_t indexr(const ModelArray::Dimensions& loc, const size_t* dims)
```

Comment:
This one assumes loc.size()=8, otherwise it will crash correct?

Reply:
Oh, yes it will. Fixed.